lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,625 @@
+  // supposed to tap in one way or another.
+  DenseMap<Attribute, BlockArgument> tappedArgs;
+  DenseMap<Attribute, Operation *> tappedOps;
+  circuitOp.walk([&](Operation *op) {
+    if (auto module = dyn_cast<FModuleOp>(op)) {
+      // Go through the module ports and collect the annotated ones.
+      for (unsigned argNum = 0, e = module.getNumArguments(); argNum < e;
+           ++argNum) {
+        auto annos = module.getArgAttrOfType<ArrayAttr>(argNum, strings.fannos);
+        if (!annos)
+          continue;
+
+        // Go through all annotations on this port and extract the interesting
+        // ones.
+        for (auto anno : annos.getAsRange<DictionaryAttr>()) {
+          auto cls = anno.getAs<StringAttr>("class");
+          if (cls == strings.referenceKeyClass)
+            assert(
+                tappedArgs.insert({anno, module.getArgument(argNum)}).second &&
+                "ambiguous tap annotation");
+        }
+      }
+    } else {
+      auto annos = op->getAttrOfType<ArrayAttr>(strings.annos);
+      if (!annos)
+        return;
+
+      // Go through all annotations on this op and extract the interesting ones.
+      // Note that the way tap annotations are scattered to their targets, we
+      // should never see multiple values or memories annotated with the exact
+      // same annotation (hence the asserts).
+      for (auto anno : annos.getAsRange<DictionaryAttr>()) {
+        auto cls = anno.getAs<StringAttr>("class");
+        if (cls == strings.memTapClass || cls == strings.referenceKeyClass ||
+            cls == strings.internalKeyClass)
+          assert(tappedOps.insert({anno, op}).second &&
+                 "ambiguous tap annotation");
+      }
+    }
+  });
+
+#ifndef NDEBUG
+  LLVM_DEBUG(llvm::dbgs() << "Tapped values:\n");
+  for (auto it : tappedArgs)
+    LLVM_DEBUG(llvm::dbgs() << "- " << it.first << ": " << it.second << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "Tapped ops:\n");
+  for (auto it : tappedOps)
+    LLVM_DEBUG(llvm::dbgs() << "- " << it.first << ": " << *it.second << "\n");
+#endif
+
+  // Process each black box independently.
+  for (auto blackBox : modules) {
+    LLVM_DEBUG(llvm::dbgs() << "Generating impls for "
+                            << blackBox.extModule.getName() << "\n");
+
+    // As a first step, gather a list of all absolute paths to instances of this
+    // black box.
+    auto paths = instanceGraph.getAbsolutePaths(blackBox.extModule);
+    for (auto path : paths)
+      LLVM_DEBUG(llvm::dbgs() << "- " << path << "\n");
+
+    // Go through the port annotations of the tap module and generate a
+    // hierarchical path for each.
+    SmallVector<PortWiring, 8> portWiring;
+    SmallDenseMap<Attribute, unsigned, 2> memPortIdx;
+    portWiring.reserve(blackBox.portAnnos.size());
+    for (auto portAnno : blackBox.portAnnos) {
+      LLVM_DEBUG(llvm::dbgs() << "- Processing port " << portAnno.portNum
```

Comment:
I'd recommend splitting this large loop body out to its method, it will allow early outs with return instead of continue and reduce nesting.

Reply:
Done, improved readability by quite a bit.